superset/assets/src/explore/components/controls/DateFilterControl.jsx

```diff
@@ -28,172 +51,265 @@ const propTypes = {
-      this.state.type = 'free';
+      this.state.type = 'range';
+      if (COMMON_TIME_FRAMES.indexOf(value) >= 0) {
+        this.state.common = value;
+      } else {
+        this.state.common = null;
+        [this.state.rel, this.state.num, this.state.grain] = value.split(' ', 3);
+      }
     }
+    this.state.isFreeform.since = !moment(this.state.since, MOMENT_FORMAT).isValid();
+    this.state.isFreeform.until = !moment(this.state.until, MOMENT_FORMAT).isValid();
+
+    // We need direct access to the state of the `DateTimeField` component
+    this.dateTimeFieldRefs = {};
   }
-  onControlChange(target, opt) {
-    this.setState({ [target]: opt.value });
-  }
-  onNumberChange(event) {
-    this.setState({ num: event.target.value });
+  onEnter(event) {
+    if (event.key === 'Enter') {
+      this.close();
+    }
   }
-  onFreeChange(event) {
-    this.setState({ free: event.target.value });
+  setStartEnd(key, value) {
+    const nextState = {
+      type: 'startend',
+      isFreeform: { ...this.state.isFreeform },
+    };
+    if (value === INVALID_DATE_MESSAGE) {
+      // the DateTimeField component will return `Invalid date` for freeform
+      // text, so we need to cheat and steal the value from the state
+      const freeformValue = this.dateTimeFieldRefs[key].state.inputValue;
+      nextState.isFreeform[key] = true;
+      nextState[key] = freeformValue;
+    } else {
+      nextState.isFreeform[key] = false;
+      nextState[key] = value;
+    }
+    this.setState(nextState, this.updateRefs);
   }
-  setType(type) {
-    this.setState({ type });
+  setCommonRange(timeFrame) {
+    const nextState = {
+      type: 'range',
+      common: timeFrame,
+      until: moment().startOf('day').format(MOMENT_FORMAT),
+    };
+    let units;
+    if (timeFrame === 'Yesterday') {
+      units = 'days';
+    } else {
+      units = timeFrame.split(' ')[1] + 's';
+    }
+    nextState.since = moment().startOf('day').subtract(1, units).format(MOMENT_FORMAT);
+    this.setState(nextState, this.updateRefs);
   }
-  setValueAndClose(val) {
-    this.setState({ type: 'free', free: val }, this.close);
+  setCustomRange(key, value) {
+    const nextState = { ...this.state, type: 'range', common: null };
+    if (key !== undefined && value !== undefined) {
+      nextState[key] = value;
+    }
+    if (nextState.rel === 'Last') {
+      nextState.until = moment().startOf('day').format(MOMENT_FORMAT);
+      nextState.since = moment()
+        .startOf('day')
+        .subtract(nextState.num, nextState.grain)
+        .format(MOMENT_FORMAT);
+    } else {
+      nextState.until = moment()
+        .startOf('day')
+        .add(nextState.num, nextState.grain)
+        .format(MOMENT_FORMAT);
+      nextState.since = moment().startOf('day').format(MOMENT_FORMAT);
+    }
+    this.setState(nextState, this.updateRefs);
   }
-  setDatetime(dttm) {
-    this.setState({ dttm: dttm.format().substring(0, 19) });
+  updateRefs() {
```

Comment:
this seems like it adds quite a bit of indirection. Would it be possible just to pass `value` as a prop to these components rather than setting their state via a ref? https://github.com/YouCanBookMe/react-datetime#api

Reply:
This is needed for freeform text. The calendar component uses `moment` to parse the string, and if it fails the state is not set. This prevents us from initializing the component saved with freeform values.